src/ucp/proto/proto_select.c

```diff
@@ -350,31 +314,33 @@ void ucp_proto_select_caps_cleanup(ucp_proto_caps_t *caps)
 static void
 ucp_proto_select_cleanup_protocols(ucp_proto_select_init_protocols_t *proto_init)
 {
-    ucp_proto_id_t proto_id;
+    ucp_proto_init_elem_t *init_elem;
 
-    ucs_for_each_bit(proto_id, proto_init->mask) {
-        ucp_proto_select_caps_cleanup(&proto_init->caps[proto_id]);
+    ucs_array_for_each(init_elem, &proto_init->protocols) {
+        ucp_proto_select_caps_cleanup(&init_elem->caps);
     }
-    ucs_free(proto_init->priv_buf);
-}
-
-static int
-ucp_proto_select_thresholds_is_last_proto(const ucp_proto_thresh_t *thresholds,
-                                          ucp_proto_id_t proto_id)
-{
-    const ucp_proto_t *proto = ucp_protocols[proto_id];
-
-    return !ucs_array_is_empty(thresholds) &&
-           (ucs_array_last(thresholds)->proto_config.proto == proto);
+    ucs_array_cleanup_dynamic(&proto_init->priv_buf);
+    ucs_array_cleanup_dynamic(&proto_init->protocols);
 }
 
-static const void *ucp_proto_select_init_priv_buf(
+static int ucp_proto_select_thresholds_is_last_proto(
         const ucp_proto_select_init_protocols_t *proto_init,
-        ucp_proto_id_t proto_id)
+        const ucp_proto_thresh_t *thresholds, ucp_proto_id_t proto_idx)
 {
-    size_t priv_offset = proto_init->priv_offsets[proto_id];
+    const void *proto_priv = ucp_proto_select_init_priv_buf(proto_init,
+                                                            proto_idx);
+    ucp_proto_config_t *last_elem;
+    ucp_proto_id_t proto_id;
+
+    if (ucs_array_is_empty(thresholds)) {
```

Comment:
Shouldn't we print warning or error in case of empty thresholds array?

Reply:
this function is removed, but if array is empty it just means the current elem is not last (because there is no last)